stan/math/gpu/opencl_context.hpp

```diff
@@ -0,0 +1,397 @@
+ */
+
+namespace stan {
+namespace math {
+
+/**
+ * The <code>opencl_context_base</code> class represents an OpenCL context
+ * in the standard Meyers singleton design pattern.
+ *
+ * See the OpenCL specification glossary for a list of terms:
+ * https://www.khronos.org/registry/OpenCL/specs/opencl-1.2.pdf.
+ * The context includes the set of devices available on the host, command
+ * queues, manages kernels.
+ *
+ * This is designed so there's only one instance running on the host.
+ *
+ * Some design decisions that may need to be addressed later:
+ * - we are assuming a single OpenCL platform. We may want to run on multiple
+ * platforms simulatenously
+ * - we are assuming a single OpenCL device. We may want to run on multiple
+ * devices simulatenously
+ */
+class opencl_context_base {
+  friend class opencl_context;
+
+ private:
+  /**
+   * Construct the opencl_context by initializing the
+   * OpenCL context, devices, command queues, and kernel
+   * groups.
+   *
+   * This constructor does the following:
+   * 1. Gets the available platforms and selects the platform
+   *  with id OPENCL_PLATFORM_ID.
+   * 2. Gets the available devices and selects the device with id
+   *  OPENCL_DEVICE_ID.
+   * 3. Creates the OpenCL context with the device.
+   * 4. Creates the OpenCL command queue for the selected device.
+   * 5. Initializes the kernel groups by filling the <code> kernel_info </code>
+   *  map.
+   * @throw std::system_error if an OpenCL error occurs.
+   */
+  opencl_context_base() {
+    try {
+      // platform
+      cl::Platform::get(&platforms_);
+      platform_ = platforms_[OPENCL_PLATFORM_ID];
+      platform_name_ = platform_.getInfo<CL_PLATFORM_NAME>();
+      platform_.getDevices(DEVICE_FILTER, &devices_);
+      if (devices_.size() == 0) {
+        system_error("OpenCL Initialization", "[Device]", -1,
+                     "CL_DEVICE_NOT_FOUND");
+      }
+      device_ = devices_[OPENCL_DEVICE_ID];
+      // context and queue
+      context_ = cl::Context(device_);
+      command_queue_ = cl::CommandQueue(context_, device_,
+                                        CL_QUEUE_PROFILING_ENABLE, nullptr);
+      // setup kernel groups
+      init_kernel_groups();
+      device_.getInfo<size_t>(CL_DEVICE_MAX_WORK_GROUP_SIZE,
+                              &max_workgroup_size_);
+    } catch (const cl::Error& e) {
+      check_opencl_error("opencl_context", e);
+    }
+  }
+  /**
+   * Initializes the <code> kernel_info </code> where each kernel is mapped to
+   * a logical flag to mark if the kernel was already compiled,
+   * the name of the kernel group, and the OpenCL kernel sources.
+   */
+  inline void init_kernel_groups() {
+    kernel_info["dummy"] = {
+        false, "timing", "__kernel void dummy(__global const int* foo) { };"};
+    kernel_info["dummy2"] = {
+        false, "timing", "__kernel void dummy2(__global const int* foo) { };"};
+  }
+  static opencl_context_base instance_;
```

Comment:
Remove this. We're using the static instance inside the static method `getInstance()`. See line 140.